### PR TITLE
Feat: support use update instead of patch for apply

### LINKED ISSUE #4469: [Feature] Support updating resources through replace
**Is your feature request related to a problem? Please describe.**

KubeVela controller now uses three-way-merge-patch to calculate patch for resource updates, and use MergePatch/StrategyMergePatch to update resources.

However, it will let the StateKeep allow merges of unintended data, like malicious sidecar being injected into pods. This is due to the patchStrategy defined by [Kubernetes API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#podspec-v1-core).

If we support using `replace` requests to update resources through specifying some policies in applications, it can help users to enforce strict consistency between KubeVela application and managed resources.

**Describe the solution you'd like**

Add a `resource-update-strategy` policy (or other name) and let the Applicator of resources to use replace requests when find this strategy. It can be configured in application level or resource level, like using `rules` in `garbage-collect` policy.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Somefive**
Signed-off-by: Somefive <yd219913@alibaba-inc.com>

### Description of your changes

Fixes #4469

I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/kubevela/kubevela/blob/master/contribute/create-pull-request.md).
- [x] [Related Docs](https://github.com/kubevela/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Added `backport release-x.y` labels to auto-backport this PR if necessary.

Notice:
1. Resources like Service have separate controller to modify its spec. Therefore, for Service resource, this configuration will not take effect.
2. KubeVela legacy rollout will not work when this flag is enabled.

### CODE REVIEW

File: pkg/utils/apply/apply.go [hunk 1]
```diff
@@ -397,3 +407,14 @@ func DryRunAll() ApplyOption {
 		return nil
 	}
 }
+
+// isUpdatableResource check whether the resource is updatable
+// Resource like v1.Service cannot unset the spec field (the ip spec is filled by service controller)
+func isUpdatableResource(desired client.Object) bool {
```
Comment (wonderflow) [inline pkg/utils/apply/apply.go:413]: I think I've written such kinds of code before, but I forget where.

File: pkg/utils/apply/apply.go [hunk 2]
```diff
@@ -170,6 +170,16 @@ func (a *APIApplicator) Apply(ctx context.Context, desired client.Object, ao ...
 		return nil
 	}
 
+	if utilfeature.DefaultMutableFeatureGate.Enabled(features.ApplyResourceByUpdate) && isUpdatableResource(desired) {
```
Comment (wonderflow) [inline pkg/utils/apply/apply.go:173]: I will use switch in this code.

switch{
case utilfeature.DefaultMutableFeatureGate.Enabled(features.ApplyResourceByUpdate) && isUpdatableResource(desired):
default:
}

Comment (Somefive) [reply pkg/utils/apply/apply.go:173]: But I think there will be no other cases here, even in the future. Is there any plan for more strategies other than Update&Patch?

Comment (Somefive) [reply pkg/utils/apply/apply.go:173]: A switch case is used in isUpdatableResource.

File: pkg/utils/apply/apply.go (hunk 1 above)
Comment (Somefive) [reply pkg/utils/apply/apply.go:413]: I think that part of code is removed. Let me find it in history archives.

Comment (Somefive) [reply pkg/utils/apply/apply.go:413]: Are you referring to https://github.com/kubevela/kubevela/blob/8de36ff2950b3c757f912e81d2fc4be3ea4b60af/pkg/controller/core.oam.dev/v1alpha2/application/dispatch/dispatch.go#L285?

File: pkg/utils/apply/apply.go (hunk 2 above)
Comment (wonderflow) [reply pkg/utils/apply/apply.go:173]: A switch will be more clear even there's only two cases

File: pkg/utils/apply/apply.go (hunk 1 above)
Comment (wonderflow) [reply pkg/utils/apply/apply.go:413]: Yes

Comment (wonderflow) [reply pkg/utils/apply/apply.go:413]: Just make sure no duplicate code

File: pkg/utils/apply/apply.go (hunk 2 above)
Comment (Somefive) [reply pkg/utils/apply/apply.go:173]: OK, agree.

File: pkg/utils/apply/apply.go (hunk 1 above)
Comment (Somefive) [reply pkg/utils/apply/apply.go:413]: No, that code only exists in release 1.1, not on master. Besides, they have different working mechanism. (Although use similar names)

**Comment by fourierr** [comment]
We will use update instead of three-way-patch later?

**Comment by Somefive** [comment]
> We will use update instead of three-way-patch later?

Currently not. Update has side-effects. It also increase the network traffic. But update can be more straightforward especially for state-keep. This feature provides an alternative usage for users who wants to use update instead of patch.

If this feature is needed in many cases, we might move it from a controller feature to an application feature or even resource feature, just like `apply-once` and `garbage-collect`. But for now, it is still only reserved for use in limited scenario.

File: pkg/utils/apply/apply.go (hunk 2 above)
Comment (Somefive) [reply pkg/utils/apply/apply.go]: Done

**Review by wonderflow — APPROVED** [review]

**Pull request merged**